nomad/structs/services.go

```diff
@@ -600,6 +601,139 @@ func (s *ConsulSidecarService) Equals(o *ConsulSidecarService) bool {
 	return s.Proxy.Equals(o.Proxy)
 }
 
+// SidecarTask represents a subset of Task fields that are able to be overridden
+// from the sidecar_task stanza
+type SidecarTask struct {
+	// Name of the task
+	Name string
+
+	// Driver is used to control which driver is used
+	Driver string
+
+	// User is used to determine which user will run the task. It defaults to
+	// the same user the Nomad client is being run as.
+	User string
+
+	// Config is provided to the driver to initialize
+	Config map[string]interface{}
+
+	// Map of environment variables to be used by the driver
+	Env map[string]string
+
+	// Resources is the resources needed by this task
+	Resources *Resources
+
+	// Meta is used to associate arbitrary metadata with this
+	// task. This is opaque to Nomad.
+	Meta map[string]string
+
+	// KillTimeout is the time between signaling a task that it will be
+	// killed and killing it.
+	KillTimeout time.Duration
+
+	// LogConfig provides configuration for log rotation
+	LogConfig *LogConfig
+
+	// ShutdownDelay is the duration of the delay between deregistering a
+	// task from Consul and sending it a signal to shutdown. See #2441
+	ShutdownDelay time.Duration
+
+	// The kill signal to use for the task. This is an optional specification,
+
+	// KillSignal is the kill signal to use for the task. This is an optional
+	// specification and defaults to SIGINT
+	KillSignal string
+}
+
+func (t *SidecarTask) Copy() *SidecarTask {
+	if t == nil {
+		return nil
+	}
+	nt := new(SidecarTask)
+	*nt = *t
+	nt.Env = helper.CopyMapStringString(nt.Env)
+
+	nt.Resources = nt.Resources.Copy()
+	nt.Meta = helper.CopyMapStringString(nt.Meta)
+
+	if i, err := copystructure.Copy(nt.Config); err != nil {
+		panic(err.Error())
+	} else {
+		nt.Config = i.(map[string]interface{})
+	}
+
+	return nt
+}
+
+// MergeIntoTask merges the SidecarTask fields over the given task
+func (t *SidecarTask) MergeIntoTask(task *Task) {
+	if t.Name != "" {
+		task.Name = t.Name
+	}
+
+	if t.Driver != "" && t.Driver != task.Driver {
```

Comment:
This if + else block is pretty subtle, but I think it makes sense. Let's add a comment so Future Us doesn't spend 10 minutes trying to remember why we did it this way.

If we ever add a new default config field which users just want to remove, this logic will bite us as it doesn't support that. I can't think of a case where setting a 0 value wouldn't accomplish the same thing, so I think this provides an optimal user experience.